Added support for write() method to submit DSA work

jcabr020 opened the pull request:
Pull request for adding support for write() method to submit DSA work and changes to page fault handling

vsanjeep commented on prov/shm/src/smr_dsa.c:
```diff
@@ -185,23 +196,94 @@ static int dsa_open_wq(struct accfg_wq *wq)
+{
+	int ret;
+	int max_comp_checks = 10000; // TODO: Should we use a different value? 
+	struct dsa_hw_desc desc = {0};
```
The descriptor must be 64 byte aligned.

vsanjeep commented on prov/shm/src/smr_dsa.c:
```diff
@@ -185,23 +196,94 @@ static int dsa_open_wq(struct accfg_wq *wq)
+static bool dsa_write_is_supported(int fd) 
+{
+	int ret;
+	int max_comp_checks = 10000; // TODO: Should we use a different value? 
```
Since this is a one time evluation, lets make it larger to 64k

vsanjeep commented on prov/shm/src/smr_dsa.c:
```diff
@@ -185,23 +196,94 @@ static int dsa_open_wq(struct accfg_wq *wq)
+	// TODO: Do we add a pause like call? 
+	while (comp.status == 0 && max_comp_checks > 0) {
+		max_comp_checks--;
+		// _mm_pause();
```
Adding __mm_pause is a good idea.

vsanjeep commented on prov/shm/src/smr_dsa.c:
```diff
@@ -185,23 +196,94 @@ static int dsa_open_wq(struct accfg_wq *wq)
+	if (fd < 0)
+		return ret;
+
+	if (*wq_mmap_support == true || *dsa_submit_cmd_mode_checked == false) {
```
This logic seems convoluted. Keep the existing function the same and handle opening the fd for write seperately in the parent function.

vsanjeep commented on prov/shm/src/smr_dsa.c:
```diff
@@ -267,10 +353,15 @@ static int dsa_idxd_init_wq_array(int shared, int numa_node,
+				&dsa_submit_cmd_mode_checked,
+				&mmapped_wq[wq_count], &wq_fd[wq_count]);
+
+			if (ret == -1)
```
Starrt with both mmap_suppot = true and wq_write_support = true. Try mapping wq with existing function with condition mmap_support == true. If it returns NULL and current wq_count==0. Set wq_mmap_support false. Write another function to open_wq for write. Try that function with wq_write_support == true. If it returns -1 and wq_count==0, set wq_write_support = false. If first time any of the worked, we use it for subsequent wq. If both fails, wq_count will remain 0.

vsanjeep replied:
And you don't need dsa_submit_cmd_mode_checked. The second function should check for wq write support. Let's do that for each wq. It is redundant but makes the logic easier to follow.

vsanjeep replied:
Actually, the second function can take a boolean value to see if write syscall check is required. The boolean value can be set as wq_count == 0 from the calling function. This way we do the check only once.

vsanjeep commented on prov/shm/src/smr_dsa.c:
```diff
@@ -903,8 +993,12 @@ void smr_dsa_context_cleanup(struct smr_ep *ep)
+	for (i = 0; i < dsa_context->wq_count; i++) {
+		if (dsa_context->wq_mmap_support)
+			dsa_idxd_wq_unmap(dsa_context->wq_portal[i]);
+		else
```
else if wq_write_supported then close the fds.

vsanjeep commented on prov/shm/src/smr_dsa.c:
```diff
@@ -148,16 +150,25 @@ static inline unsigned char dsa_enqcmd(struct dsa_hw_desc *desc,
 				dsa_context->wq_portal[dsa_context->next_wq]);
-		dsa_context->next_wq = (dsa_context->next_wq + 1) %
-			(dsa_context->wq_count);
+		else {
```
else if wq_write_supported. The the final else is an assert since as of now we  should not enter this flow if either one is not supported.

jcabr020 committed:
prov/shm: Added support for write() method to submit DSA work

DSA driver disabled mapping workqueues to user address space for some
products in order to address errata. The DSA driver instead provides write
system call as an alternative method to submit work. This commit adds the
support to use the write system call.

Signed-off-by: Jorge Cabrera <jorge.cabrera@intel.com>

jcabr020 committed:
prov/shm: Touch all buffer pages after DSA page fault

DSA page faults will now always trigger touching all of the pages in the
descriptor buffer (as it is done in dsa_touch_buffer_pages()), instead of
just the individual page causing the fault (as was done in
dsa_handle_page_fault()).

Previously, the default behavior was to touch the individual faulting page,
and if the user enabled the FI_SHM_ENABLE_DSA_PAGE_TOUCH flag, then all the
rest of the pages in the descriptor buffer were also touched. This commit
removes this flag, and now always touches all the pages instead.

Signed-off-by: Jorge Cabrera <jorge.cabrera@intel.com>

vsanjeep commented on prov/shm/src/smr_dsa.c:
```diff
@@ -132,32 +142,41 @@ struct dsa_ops {
+			&dsa_context->wq_handle[dsa_context->next_wq], hw);
+		dsa_context->next_wq =
+			(dsa_context->next_wq + 1) % (dsa_context->wq_count);
+	} while (status);
```
enq_cmd returns 1 as status when the status is retry. write_cmd function as written returns -1 as status when the status is retry. Not sure if the loop will work correctly if status is -1. Change the write_cmd function to also return 1 in case of retry.
